cedar-lean/Cedar/Spec/PartialRequest.lean

```diff
@@ -0,0 +1,62 @@
+/-
+ Copyright 2022-2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      https://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+-/
+
+import Cedar.Spec.Expr
+import Cedar.Spec.PartialExpr
+import Cedar.Spec.PartialValue
+import Cedar.Spec.Request
+import Cedar.Spec.Value
+
+/-!
+This file defines Cedar partial requests.
+-/
+
+namespace Cedar.Spec
+
+open Cedar.Data
+
+inductive UidOrUnknown where
+  | known (uid : EntityUID)
+  | unknown (name : String)
+
+deriving instance Repr, DecidableEq, Inhabited for UidOrUnknown
+
+instance : Coe UidOrUnknown PartialValue where
+  coe x := match x with
+  | .known uid    => .value uid
+  | .unknown name => .residual (PartialExpr.unknown name)
+
+structure PartialRequest where
+  principal : UidOrUnknown
+  action : UidOrUnknown
+  resource : UidOrUnknown
+  context : Map Attr RestrictedPartialValue -- allows individual context attributes to have unknown values, but does not allow it to be unknown whether a context attribute exists at all
```

Comment:
I don't follow this comment:  how is the use of RestrictedPartialValue (vs PartialValue) related to the presence of an attribute/value binding in this map?

(Or maybe "attribute exists" here doesn't refer to the map's contents?  If so, what does it refer to?)

Reply:
The comment is not about the use of RestrictedPartialValue vs PartialValue.  It is generally about the definition as `Map Attr RestrictedPartialValue` -- that is, each `Attr` is either in the map or not, and if it is in the map, it definitely has a (partial) value.  An alternate definition that would allow attributes to have unknown existence would be something like `Map Attr (Option RestrictedPartialValue)`, or something isomorphic